File: UnityiOSPlugin/src/Frontend/Notify/RemoteNotifierPlugin.swift
// ======================================================================
// Project Name    : unity plugin
//
// Copyright © 2017 U-CREATES. All rights reserved.
//
// This source code is the property of U-CREATES.
// If such findings are accepted at any time.
// We hope the tips and helpful in developing.
// ======================================================================
import Foundation
import UIKit
import UserNotifications
open class RemoteNotifierPlugin: BaseNotifierPlugin, UNUserNotificationCenterDelegate {
    fileprivate static var instance: RemoteNotifierPlugin?
    fileprivate override init() {}
    @objc
    open static func getInstance() -> RemoteNotifierPlugin! {
        if (nil == RemoteNotifierPlugin.instance) {
            RemoteNotifierPlugin.instance = RemoteNotifierPlugin()
        }
        return RemoteNotifierPlugin.instance!
    }
    @objc
    open override func register() -> Void {
        if #available(iOS 10.0, *) {
            func callback(granted: Bool, error: Error?) -> Void {
                if (nil != error) {
                    print(TagPlugin.UNITY_PLUGIN_IDENTIFIER + error.debugDescription)
                } else {
                    let app: UIApplication = UIApplication.shared
                    app.registerForRemoteNotifications()
                }
                return
            }
            let userNotificationCenter: UNUserNotificationCenter = UNUserNotificationCenter.current()
            userNotificationCenter.delegate = self
            userNotificationCenter.requestAuthorization(options: [.alert, .badge, .sound], completionHandler: callback)
        } else {
            let settings: UIUserNotificationSettings = UIUserNotificationSettings(types: [.alert, .badge, .sound], categories: nil)
            let app: UIApplication = UIApplication.shared
            app.registerUserNotificationSettings(settings)
        }
        return
    }
    @available(iOS 10.0, *)
    public func userNotificationCenter(_ center: UNUserNotificationCenter, willPresent notification: UNNotification, withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
        print(TagPlugin.UNITY_PLUGIN_IDENTIFIER + "execute remote notification.")
        let userInfo: [AnyHashable: Any] = notification.request.content.userInfo
        self.noitfy(userInfo: userInfo)
        self.reset()
        completionHandler([.alert, .badge, .sound])
        return
    }
    @available(iOS 10.0, *)
    public func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Swift.Void) {
        print(TagPlugin.UNITY_PLUGIN_IDENTIFIER + "press remote notification message in notify area.")
        let userInfo: [AnyHashable: Any] = response.notification.request.content.userInfo
        self.noitfy(userInfo: userInfo)
        self.reset()
        completionHandler()
        return
    }
}

File: UnityAndroidPlugin/plugin/src/main/java/com/frontend/notify/RemoteNotifierPlugin.java
// ======================================================================
// Project Name    : unity plugin
//
// Copyright © 2017 U-CREATES. All rights reserved.
//
// This source code is the property of U-CREATES.
// If such findings are accepted at any time.
// We hope the tips and helpful in developing.
// ======================================================================
package com.frontend.notify;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.util.Log;
import android.view.Window;
import android.view.WindowManager;
import com.core.identifier.TagPlugin;
import com.frontend.activity.ActivityPlugin;
import com.google.android.gms.gcm.GoogleCloudMessaging;
import java.io.IOException;
public class RemoteNotifierPlugin extends BaseNotifierPlugin {
    private static RemoteNotifierPlugin instance = null;
    private RemoteNotifierPlugin() {}
    public static RemoteNotifierPlugin getInstance() {
        if (null == RemoteNotifierPlugin.instance) {
            RemoteNotifierPlugin.instance = new RemoteNotifierPlugin();
        }
        return RemoteNotifierPlugin.instance;
    }
    public void register(final String senderId) {
        final Activity activity = ActivityPlugin.getInstance();
        final Context context = activity.getApplicationContext();
        AsyncTask<Void, Void, String> task = new AsyncTask<Void, Void, String>() {
            @Override
            protected String doInBackground(Void... params) {
                String gcmToken = "";
                try {
                    GoogleCloudMessaging gcm = GoogleCloudMessaging.getInstance(context);
                    gcmToken = gcm.register(senderId);
                    enable = true;
                    Log.i(TagPlugin.UNITY_PLUGIN_IDENTIFIER, "GCMToken::" + gcmToken);
                } catch (IOException e) {
                    enable = false;
                    Log.i(TagPlugin.UNITY_PLUGIN_IDENTIFIER, e.getMessage());
                }
                return gcmToken;
            }
            @Override
            protected void onPostExecute(String msg) {
                Log.i(TagPlugin.UNITY_PLUGIN_IDENTIFIER, msg);
                Window window = activity.getWindow();
                window.setFlags(WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN, WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN);
                return;
            }
        };
        task.execute(null, null, null);
        return;
    }
    @Override
    public void register(Context context, Intent intent) {
        String udt = intent.getStringExtra("google.c.a.udt");
        if (null == udt) {
            this.enable = true;
        }
        this.context = context;
        return;
    }
}
